middleware/proxy/dns.go

```diff
@@ -109,7 +91,7 @@ func exchange(m *dns.Msg, co net.Conn) (dns.Msg, error) {
 
 	dnsco.Close()
 	if r == nil {
-		return dns.Msg{}, err
+		return nil, err
```

Comment:
The return value was changed back to a pointer, `exchange` was only used from within the single inflight construct so any changes would be very localized.